### PR TITLE
feat(stepfunctions-tasks): runtime role in EmrAddStep 

### LINKED ISSUE #27691: aws-stepfunctions-tasks: Add support for ExecutionRoleArn to EmrAddStep
### Describe the feature

On October 22, 2022, [EMR launched the Runtime Roles feature](https://aws.amazon.com/blogs/big-data/introducing-runtime-roles-for-amazon-emr-steps-use-iam-roles-and-aws-lake-formation-for-access-control-with-amazon-emr/) to allow jobs to execute as a more specific role than the cluster.  This additionally opened up the ability to utilize LakeFormation to access data shared to your job's execution role.

This feature added a new, optional parameter named [`ExecutionRoleArn` to the `AddJobFlowSteps`](https://docs.aws.amazon.com/emr/latest/APIReference/API_AddJobFlowSteps.html#EMR-AddJobFlowSteps-request-ExecutionRoleArn) action.  Consequently, the matching StepFunction action of [`addStep` and `addStep.sync` have also added this optional parameter](https://docs.aws.amazon.com/step-functions/latest/dg/connect-emr.html).

> Optionally, you can also specify the [ExecutionRoleArn](https://docs.aws.amazon.com/emr/latest/APIReference/API_AddJobFlowSteps.html#EMR-AddJobFlowSteps-request-ExecutionRoleArn) parameter while using this API.

I'd like to have `EmrAddStep` support this new `ExecutionRoleArn` field so that I can utilize the Runtime Roles feature on clusters which are managed by a StepFunctions state machine.

### Use Case

I specifically intend to use this functionality to migrate an existing process to utilize LakeFormation's access delegation instead of having an instance role which has to have full access to the underlying S3 bucket.

### Proposed Solution

In order to implement, we only need to add some property to the step's props then pass that value through when rendering the task.  There are two reasonable options that I see:

### Option 1 - Expose an `executionRoleArn` property as a `string`
In order to keep the solution as simple as possible and avoid the same issue as #21319, we can simply expose the parameter as an optional string.

1. Add a new `executionRole` parameter to the `EmrAddStepProps` interface:
   ```
   export interface EmrContainersStartJobRunProps extends sfn.TaskStateBaseProps {
     ...
     readonly executionRoleArn?: string;
     ...
   }
   ```
2. Update `_renderTask()` to emit the required `ExecutionRoleArn` field when it is provided:
   ```
     protected _renderTask(): any {
       return {
         ...
         Parameters: sfn.FieldUtils.renderObject({
           ...
           ExecutionRoleArn: this.props.executionRoleArn,
           ...
         }),
       };
     }
   ````

### Option 2 - Expose `executionRole` as an `IRole` and `executionRoleArn` as a `string`
In #21319, it appears that we had originally only implemented exposing a `executionRole` as `IRole` and only later realized that doesn't work for JsonPath-provided values.  If we want to stay consistent with that pattern, we can do the same.

1. Add a new `executionRole` parameter to the `EmrAddStepProps` interface:
   ```
   export interface EmrContainersStartJobRunProps extends sfn.TaskStateBaseProps {
     ...
     readonly executionRole?: iam.IRole;
     readonly executionRoleArn?: string;
     ...
   }
   ```
2. Validate that only either `executionRole` or `executionRoleArn` are provided:
   ```
   constructor(scope: Construct, id: string, private readonly props: EmrAddStepProps) {
     ...
     if (props.executionRole !== undefined && props.executionRoleArn !== undefined) {
       throw new Error(...);
     }
     ...
   }
   ```
3. Update `_renderTask()` to emit the required `ExecutionRoleArn` field when it is provided:
   ```
     protected _renderTask(): any {
       return {
         ...
         Parameters: sfn.FieldUtils.renderObject({
           ...
           ExecutionRoleArn: this.props.executionRoleArn ?? this.props.executionRole?.roleArn,
           ...
         }),
       };
     }
   ````

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.95.1

### Environment details (OS name and version, etc.)

Mac 13.5

### PR DESCRIPTION
**Pull request by msambol**
Here is the result of running the step function deployed by the integration test. The step completed successfully with the IAM role.

emr_runtime_role

Closes #27691.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:22]: Given the prerequisites required and resource cost of integration test, I commented out some of this. The step function deploys successfully and I manually deployed the stack and ran the step function to confirm.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts:100]: I'd prefer this be `executionRoleArn` but the linter didn't like it because this isn't an L1 construct.

**Comment by msambol** [comment]
@mattliemAWS if you're interested.

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (lpizzinidev) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:14]:
```suggestion
 * Prerequisites:
 * 1. Fill in the instances block for EmrCreateCluster
 * 2. Create the EMR security configuration, see https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-create-security-configuration.html
 * 3. Add the security configuration to EmrCreateCluster
```

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (lpizzinidev) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:97]:
```suggestion

const executionRolePrincipal = new iam.ArnPrincipal(createClusterStep.clusterRole.roleArn);
const executionRole = new iam.Role(stack, 'Role', {
  roleName: 'EmrStepExecutionRole',
  assumedBy: executionRolePrincipal,
});
executionRole.assumeRolePolicy?.addStatements(
    new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      principals: [
        executionRolePrincipal,
      ],
      actions: [
        'sts:SetSourceIdentity',
      ],
    }),
    new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      principals: [
        executionRolePrincipal,
      ],
      actions: [
        'sts:TagSession',
      ],
      conditions: {
        StringEquals: {
          'aws:RequestTag/LakeFormationAuthorizedCaller': 'Amazon EMR',
        },
      },
    }),
  );
```
This should generate the necessary policy.

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (lpizzinidev) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:41]: Is the `ec2SubnetId` property necessary?
Can't we just remove the comments and give a default configuration?

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks!
The implementation looks good.
Some adjustments are needed on documentation and tests in my opinion.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts
Comment (msambol) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts:100]: It doesn't like anything that ends with `arn` because of [this line](@see https://aws.amazon.com/blogs/big-data/introducing-runtime-roles-for-amazon-emr-steps-use-iam-roles-and-aws-lake-formation-for-access-control-with-amazon-emr/). I could use `eslint-disable` ?

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (msambol) [reply packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:41]: I couldn't get it to deploy without it. I kept getting `the subnet id '' does not exist`... maybe a bug honestly.

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (lpizzinidev) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:43]:
```suggestion
const vpc = new ec2.Vpc(stack, 'Vpc', { restrictDefaultSecurityGroup: false });

const createClusterStep = new tasks.EmrCreateCluster(stack, 'EmrCreateCluster', {
  instances: {
    instanceFleets: [
      {
        instanceFleetType: tasks.EmrCreateCluster.InstanceRoleType.MASTER,
        instanceTypeConfigs: [
          {
            instanceType: 'm5.xlarge',
          },
        ],
        targetOnDemandCapacity: 1,
      },
    ],
    ec2SubnetId: vpc.publicSubnets[0].subnetId,
  },
```
Let's provide a comprehensive example, integration tests can be a useful template for users who want to implement the functionality (may need adjustments).
Creating a cluster from the console requires a VPC and subnet(s), I wasn't able to find documentation about it.

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (lpizzinidev) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:14]:
```suggestion
 * Create and add an EMR security configuration to EmrCreateCluster.
 * See https://aws.amazon.com/blogs/big-data/introducing-runtime-roles-for-amazon-emr-steps-use-iam-roles-and-aws-lake-formation-for-access-control-with-amazon-emr/
```

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts:98]:
```suggestion
   * @default - Uses EC2 instance profile role
```

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks for the follow-up.
I think that we should automate as much as possible on the integration test to make it maintainable and re-usable as a reference.
Some smaller adjustments are needed on documentation.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts
Comment (lpizzinidev) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/emr/emr-add-step.ts:100]:
```suggestion
  readonly executionRoleArn?: string;
```
After maintainer feedback, let's rename this `executionRoleArn` and add the lint exception.

File: packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:42]: @lpizzinidev I added this as well so no prereqs are required. Step function succeeds after deploying stack as-is. Good shout.

Comment (lpizzinidev) [reply packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/emr/integ.emr-create-cluster-add-step-runtime-role.ts:42]: Nice 👍

**Review by lpizzinidev — APPROVED** [review]
Thanks 💪

**Comment by vinayak-kukreja** [comment]
Hey @msambol , thank you for this contribution. I am seeing the integ test fails for `EmrCreateCluster`. I am seeing,

```
        "StateChangeReason": {
          "Code": "VALIDATION_ERROR",
          "Message": "EMR service role arn:aws:iam::<AccountId>:role/aws-cdk-emr-add-step-runt-EmrCreateClusterServiceRo-xZEbpPmy7it3 is invalid"
        },
```
Screenshot 2023-12-05 at 3 54 41 PM

Could you confirm if its working on your end once again?

**Comment by msambol** [comment]
> Hey @msambol , thank you for this contribution. I am seeing the integ test fails for `EmrCreateCluster`. I am seeing,
> 
> ```
>         "StateChangeReason": {
>           "Code": "VALIDATION_ERROR",
>           "Message": "EMR service role arn:aws:iam::<AccountId>:role/aws-cdk-emr-add-step-runt-EmrCreateClusterServiceRo-xZEbpPmy7it3 is invalid"
>         },
> ```
> 
> Screenshot 2023-12-05 at 3 54 41 PM
> Could you confirm if its working on your end once again?

Can I see your code?

**Comment by msambol** [comment]
It was working. I'll take a look this weekend.

**Comment by msambol** [comment]
@vinayak-kukreja I think there is a bug with the EMR create cluster process right now. It won't let me create a cluster with a perfectly-acceptable EMR service role. I'm checking with the EMR team.

**Comment by vinayak-kukreja** [comment]
> Can I see your code?

Hey @msambol , I just deployed the integ test mentioned in this PR. The deployment succeeds but when I run the first command mentioned in steps within the integ test, it fails the create task. Let me know if I can provide any other information.

**Comment by msambol** [comment]
Mine failed as well. Something is off with EMR cluster creation. I reached out to the EMR team.

**Comment by vinayak-kukreja** [comment]
Ok, thank you. Let me know if you hear anything from them. If not, I can reach out too.

**Comment by msambol** [comment]
@vinayak-kukreja I addressed the cluster creation issues in https://github.com/aws/aws-cdk/pull/28327. Once that's merged, I'll rebase and make sure this integration test succeeds.

**Comment by paulhcsun** [comment]
Hi @msambol, your fix for ecr cluster creation https://github.com/aws/aws-cdk/pull/28327 has been approved and merged.

**Comment by msambol** [comment]
@paulhcsun Rebasing now and re-running integ test 👍🏼

**Comment by msambol** [comment]
@paulhcsun This has been updated and confirmed working.

**Review by paulhcsun — APPROVED** [review]
Thanks for updating and confirming the tests are working. Great work and thanks for the contribution @msambol!

**Pull request merged**
